platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+          ErrorCode.UNKNOWN, 'testing')
+      '2 attempts where the first one failed with IOException'         || IOException
+      '2 attempts where the first one failed with RutimeException'     || RuntimeException
+  }
+
+  @Unroll
+  // the following needs to be changed as it should be closing the whole thing and not re-execute the creation
+  def 'test constructor when client is created after 2 attempts where the first one is interrupted with #exception.simpleName and becomes available'() {
+    given: "a client and creator to verify that failsafe actually delegates to our creator and client"
+      def client = Mock(SolrClient) {
+        1 * ping() >> pingOkResponse
+      }
+      def creator = Mock(Creator) {
+        1 * create() >> client
+      }
+
+    and: "controllers that will actually proceed through Solr"
+      def createController = new FailsafeController('SolrClient Creation')
+          .onNextExecution(doThrow(exception))
+          .and().onNextExecution(doProceed())
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution(doProceed())
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the reported core should correspond to the requested one"
+      adapter.core == CORE
+
+    and:
+      "our creator should be registered with the adapter"
+      adapter.creator.is(creator)
+
+    when: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.isAvailable()
+
+    and: "the underlying client should be the one we created"
+      adapter.proxiedClient.is(client)
+
+    and: "the underlying client should never be closed"
+      0 * client.close()
+
+    and: "there should be no registered initializers or listeners"
+      !adapter.hasInitializers() && !adapter.hasListeners()
+
+    and: "verify failsafe controllers"
+      createController.verify()
+      pingController.verify()
+
+    cleanup:
+      createController?.shutdown()
+      pingController?.shutdown()
+
+    where:
+      exception << [InterruptedException, InterruptedIOException]
+  }
+
+  @Unroll
+  def 'test client #test_client_how where an API call also fails to connect after the initial failure'() {
+    given: "a Solr client that fails direct pings"
+      def client = Mock(SolrClient) {
+        // in case the real client is consulted for the first API call
+        // since either we will be rejected right away if the client is not created but if it is
+        // a ping will be issued to the client
+        ping() >> pingFailResponse
+      }
+
+    and: "controllers that mocks every execution/attempts"
+      def createController = new FailsafeController('SolrClient Creation').onNextExecution {
+        waitTo('create').onlyIf(blocking_create)
```

Comment:
both are aliases but I think in this case `wait to create` reads better than `wait for create` since the next step in this mock is to create the client.

As for the `onlyIf()`, I agree with you. In this case, since it is fairly concentrate on one or 2 things and they should be clear enough that I am on the fence. I did add a comment to warn developers to not take this as a good thing to do all the time